File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
def solution(s):
    answer = len(s)
    
    if answer == 1:
        return 1
    
    for i in range(1, len(s)//2 + 1):
        compressed = ''
        prev = s[0:i]
        cnt = 1
        for j in range(i, len(s), i):
            next = s[j:j+i]
            if prev == next:
                cnt += 1
            else:
                if cnt >= 2:
                    compressed += str(cnt) + prev
                else:
                    compressed += prev
                prev = next
                cnt = 1

        if cnt >= 2:
                    compressed += str(cnt) + prev
        else:
            compressed += prev
        answer = min(answer, len(compressed))
    
    return answer

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
class Solution {
    public int solution(String s) {
        int answer = Integer.MAX_VALUE;

        int maxUnitLength = s.length() / 2;
        for (int unitLength = 1; unitLength <= maxUnitLength; unitLength++) {
            String unit = s.substring(0, unitLength);
            int cnt = 0;
            StringBuilder sb = new StringBuilder();
            int lastIndex = 0;
            while (lastIndex <= s.length() - unitLength) {
                String str = s.substring(lastIndex, lastIndex + unitLength);
                if (str.equals(unit)) {
                    cnt++;
                } else {
                    if (cnt == 1) {
                        sb.append(unit);
                    } else {
                        sb.append(cnt).append(unit);
                    }
                    cnt = 1;
                    unit = str;
                }
                lastIndex += unitLength;
            }
            if (cnt == 1) {
                sb.append(unit);
            } else {
                sb.append(cnt).append(unit);
            }
            sb.append(s.substring(lastIndex));
            answer = Math.min(answer, sb.length());
        }

        if (answer == Integer.MAX_VALUE) {
            return 1;
        }

        return answer;
    }
}